Add EIP: Add Controlled Gas Limit Increase Strategy (#8933)
ethereum/EIPs: EIPS/eip-7783.md

## **Abstract**

The EIP proposes the introduction of a new gas limit management mechanism that automatically increases the block gas limit over time. The incremental growth is controlled by a fixed rate, ensuring predictable network scaling while preventing sudden surges in block sizes. This strategy is meant to be used as a default setting, with the option to switch to a fixed gas limit if needed (or different parameters).

## **Motivation**

### **Predictable Gas Limit Growth**

- **Current Issue:**
  - The Ethereum network faces increasing demand, but changes to the gas limit are often manually adjusted by miners or validators based on their preferences, which may cause unpredictable block sizes and network performance issues.

- **Need for Change:**
  - A systematic and predictable increase of the gas limit will help scale the network while giving the ecosystem time to adjust to larger block sizes, without needing to rely on ad hoc decisions by network participants.

### **Gradual Increase with Deactivation Safeguard**

- **Controlled Growth:**

>   - Instead of sudden or unpredictable changes, this EIP proposes incremental gas Limit increases over a specified amount of time, ensuring a smooth transition to higher transaction throughput, while still keeping the governance of the gas Limit in the hand of the community.

Reviewer: ```suggestion
  - Instead of sudden or unpredictable changes, this EIP proposes incremental gas limit increases over a specified amount of time, ensuring a smooth transition to higher transaction throughput, while still keeping the governance of the gas limit in the hand of the community.
```

  
- **Automatic Deactivation:**
  - A safeguard deactivation block will halt the increase after some specified time, preventing the gas Target from growing indefinitely and allowing the community to reassess the network's needs.

## **Specification**

### **Incremental Gas Target Increase Strategy**

Add a new "Gas Target" selection strategy that takes in Block Number `N` and spits out the Gas Target `GT` for that block. The strategy is as follows:

 - The gas target `GT_t` at block `t` is calculated as:

```python

> def compute_gas_target(blockNum: int, blockNumStart: int, initialGasTarget: int, r: int, gasTargetCap: int) -> int:

Reviewer 2: and gasLimit stays same or limit also grows with the 2X relation?

Reviewer 2: if gasLimit is also changing in same proportion, then `2r` has to be less < gasLimit/1024 which is easily statisfied i guess, but still needs a mention for context of `r`

Author: yes - if the gas target is increased in a large amount sample of validators, yes

Author: I mentioned that gas limit still grows 2x in proportion to the gas target. tell me if it is satisfiable

  if blockNum < blockNumStart:
    return initialGasLimit
  else:
    return min(gasLimitCap, initialGasLimit + r * (blockNum - blockNumStart))
```

Where:

- `blockNum` is the block number for which the gas limit is being calculated.
- `blockNumStart` is the block number at which the gas limit increase starts.
- `initialGasLimit` is the initial gas limit at block `blockNumStart`.
- `r` is the rate at which the gas limit increases per block.
- `gasLimitCap` is the maximum gas limit that can be reached.

> If we set `blockNumStart` to the current block number, `initialGasLimit` to the current gas limit (`30_000_000`), `r` to 6, and `gasLimitCap` to `60_000_000`, the gas limit will increase by 3 gas per block for two years, reaching 30 million gas at the end of a 2 years period. 

Reviewer 2: ```suggestion
If we set `blockNumStart` to the current block number, `initialGasLimit` to the current gas limit (`30_000_000`), `r` to 6, and `gasLimitCap` to `60_000_000`, the gas limit will increase by 6 gas per block for two years, reaching 30 million gas at the end of a 2 years period. 
```
although approved, can you correct this, just noticed

> If we set `blockNumStart` to the current block number, `initialGasLimit` to the current gas limit (`30_000_000`), `r` to 6, and `gasLimitCap` to `60_000_000`, the gas limit will increase by 6 gas per block for two years, reaching 30 million gas at the end of a 2 years period. 

Reviewer: Nit: note that this is ~694 days (~1.9 years). Caluculation: 

```
blocksNecessary = (30_000_000 / 6)
timeNeeded = blocksNecessary * 12
days = timeNeeded / (24 * 60  * 60)
```

This assumes 12s block time.

> 

Reviewer: I would note somewhere that this formula is the current gas limit which we "want". The gas limit is still subject to certain checks, namely: the proposed gas limit should stay within bounds of the consensus rules. I think, but have to check, that currently the gas limit can only decrease / increase by 1/1024 of the parent blocks' gas limit.

### **Implementation**

- **Client Updates:**

>   - Ethereum clients must implement the automatic gas limit increase according to the specified rate and formula, with enforcement of the deactivation block. However, If the GasLimit has been locally set by the validator that will override the formula and take precedence.

Reviewer: ```suggestion
  - Ethereum clients must implement the automatic gas limit increase according to the specified rate and formula, with enforcement of the deactivation block. However, if the gas limit has been locally set by the validator that will override the formula and take precedence.
```

I also find the second sentence "However..." formatted a bit weird :sweat_smile:

  
- **Gas Limit Management:**

>   - Miners and validators are not required to adjust the gas limit manually during this period, as the increase will be automated at the protocol level.

Reviewer: I do not understand this, the gas limit formula is optional, so nothing is enforced by the protocol. What is meant by this sentence?

Author: ooops - sorry it is a left over

## **Rationale**

### **Predictable Growth**

- **Systematic Adjustment:**
  - The gradual increase avoids sudden surges in gas target that could destabilize the network. Instead, it provides a smooth transition, giving the ecosystem time to adapt to larger block sizes.

### **Controlled Target with Deactivation Block**

- **Automatic Safeguard:**
  - The inclusion of a deactivation block ensures that the gas target does not increase indefinitely, preventing potential negative impacts on network performance beyond the planned growth.

- **Community Consensus:**
  - The deactivation block serves only as a checkpoint for the community to evaluate the impact of the gas target increase, however, in the two-year period, the community can decide to halt the increase at any time and can also switch to a fixed gas target if needed.

## **Backwards Compatibility**

**No Hard Fork Required**

## **Security Considerations**

- The controlled gas target increase strategy is designed to prevent sudden changes that could lead to network instability or security vulnerabilities.

> - The fact that validators can re-adjust the gas target in case of a DDOS attack or other issues, makes the network more secure than to just increase the gas target manually in a cliff-like manner.

Reviewer 3: ```suggestion
- The fact that validators can re-adjust the gas target in case of a DOS attack or other issues, makes the network more secure than to just increasing the gas target manually in a cliff-like manner.
```
